internal/stacks/stackruntime/internal/stackeval/component_config.go

```diff
@@ -422,12 +431,24 @@ func (c *ComponentConfig) ExprReferenceValue(ctx context.Context, phase EvalPhas
 	return cty.DynamicVal
 }
 
-func (c *ComponentConfig) ResolveExpressionReference(ctx context.Context, ref stackaddrs.Reference) (Referenceable, tfdiags.Diagnostics) {
+func (c *ComponentConfig) ResolveExpressionReference(ctx context.Context, phase EvalPhase, ref stackaddrs.Reference) (Referenceable, tfdiags.Diagnostics) {
 	repetition := instances.RepetitionData{}
 	if c.Declaration(ctx).ForEach != nil {
+
+		forEachValue := c.ForEachValue(ctx, phase)
```

Comment:
As far as I remember, this is the first example of one of the "-Config" types trying to resolve `for_each`. That doesn't necessarily mean it's wrong of course, but it does give me pause if only because it's making things inconsistent: we aren't doing the same thing for `StackCallConfig` or `ProviderConfig` (as far as I know!), and we're evaluating in the scope of a `StackConfig` rather than a `Stack` so the expression evaluation context is gonna be an interesting mixture of known and unknown values at this stage.

In a world where we can better tolerate unknown values in `for_each` (i.e. with "deferred actions") this might be fine. It does have the nice advantage of being able to return a more specific type constraint on the unknown value, and so I'm not meaning to say that I'm opposed to it, only that I worry that treating this one case differently than the others might cause confusing and inconsistent results.

Are you intending to make similar changes for the other types that represent configuration objects that can have `for_each`?

Reply:
Yeah, that's fair and I hadn't thought about the others. Given I had a lot of extra changes adding the Phase into the function signature, I've rolled this back for now. It's not essential, and would be good to maintain consistency. We can come back and add this again later if we decide we need it for all three instead of just the single one 👍